src/klein/_app.py

```diff
@@ -53,11 +80,28 @@ def _call(__klein_instance__, __klein_f__, *args, **kwargs):
 @implementer(IKleinRequest)
 class KleinRequest(object):
     def __init__(self, request):
+        # type: (Request) -> None
         self.branch_segments = [""]
-        self.mapper = None
 
-    def url_for(self, *args, **kwargs):
-        return self.mapper.build(*args, **kwargs)
+        # Don't annotate as optional, since you should never set this to None
+        self.mapper = None  # type: MapAdapter # type: ignore[assignment]
+
+    def url_for(
+        self,
+        endpoint,  # type: Text
+        values=None,  # type: Optional[Mapping[Text, Text]]
+        method=None,  # type: Optional[Text]
+        force_external=False,  # type: bool
+        append_unknown=True,  # type: bool
+    ):
+        # type: (...) -> Text
+        assert self.mapper is not None
```

Comment:
I don't think this `assert` is required any more, given that you've got the `type: MapAdapter` above.

If you want to leave it in for clarity, please add an assertion message explaining what the "right" way to construct a KleinRequest is...

Reply:
0959687366bcd53161e4facb03485aee31993ae9